pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -182,3 +300,101 @@ func (w *HashAggPartialWorker) shuffleIntermData(finalConcurrency int) {
 		w.outputChs[i] <- &w.partialResultsMap[i]
 	}
 }
+
+func (w *HashAggPartialWorker) prepareForSpill() {
+	if !w.isSpillPrepared {
+		w.tmpChksForSpill = make([]*chunk.Chunk, spilledPartitionNum)
+		w.spilledChunksIO = make([]*chunk.DataInDiskByChunks, spilledPartitionNum)
+		for i := 0; i < spilledPartitionNum; i++ {
+			w.tmpChksForSpill[i] = w.spillHelper.getNewSpillChunkFunc()
+			w.spilledChunksIO[i] = chunk.NewDataInDiskByChunks(w.spillHelper.spillChunkFieldTypes)
+			if w.spillHelper.diskTracker != nil {
+				w.spilledChunksIO[i].GetDiskTracker().AttachTo(w.spillHelper.diskTracker)
+			}
+		}
+		w.isSpillPrepared = true
+	}
+}
+
+func (w *HashAggPartialWorker) spillDataToDisk() error {
+	err := w.spillDataToDiskImpl()
+	if err == nil {
+		err = failpointError()
+	}
+	return err
+}
+
+func (w *HashAggPartialWorker) spillDataToDiskImpl() error {
+	defer func() {
+		if r := recover(); r != nil {
+			recoveryHashAgg(w.globalOutputCh, r)
+		}
+	}()
+	if len(w.partialResultsMap) == 0 {
```

Comment:
Move this if to line 328 and merge the 2 defers?

Reply:
> Move this if to line 328 and merge the 2 defers?

okk, but I think we can move it to L341.